sdk-typed-search-attributes.md

```diff
@@ -0,0 +1,497 @@
+* Python
+  * Search attributes getting in workflow - on info as `Mapping[str, List[scalar]]`
+  * Search attributes setting in workflow - via `upsert_search_attributes` accepting `Mapping[str, List[scalar]]`
+  * Search attributes setting on start - via `Mapping[str, List[scalar]]` on start options
+  * Search attributes getting on list/describe - as `Mapping[str, List[scalar]]`
+* .NET
+  * Still in alpha, can decide whatever we want
+
+### Server
+
+* ElasticSearch
+  * Multi-value search attribute for non-keyword-list - accepted, but discouraged, and returned as given
+    * TODO(cretz): Question outstanding to server team about the benefits of multi-value non-keyword-list attributes
+  * Single-value search attribute for non-keyword-list - accepted and returned as given
+  * Multi-value search attribute for keyword-list - accepted and returned as given
+  * Single-value search attribute for keyword-list - fails
+* SQL
+  * Multi-value search attribute for non-keyword-list - only allowed if a single value, always returned single value
+  * Single-value search attribute for non-keyword-list - accepted and returned as given
+  * Multi-value search attribute for keyword-list - accepted and returned as given
+  * Single-value search attribute for keyword-list - fails
+
+### Problems with current behavior
+
+* Go and Java rely on user to resolve ambiguity between single and multi value search attributes (Go is worse because it has no helpers)
+* TypeScript and Python always disambiguate making all search attributes a list (and Java too in single-value cases) even if list not allowed anymore
+* Setting invalid search attributes in workflow never returns a failure that can be handled by workflow code
+* Users have no way to clarify the difference between keyword and keyword-list
+
+## Goals
+
+Immediate:
+
+* As much as is supported by the language, users need to be able to get and set individually typed search attributes
+* Maintain backwards compatibility but deprecate any approach that doesn't let you identify the search attribute type
+  alongside the use of its value
+
+Non-immediate:
+
+* Provide ways to define predefined search attributes that are typed
+* Also provide ways to define predefined memo values that are typed
+
+Non-goals:
+
+* Better validation of actual search attribute type from server
+* Better way for server to tell SDK the upsert failed
+* Support for user-explicit upsert of multiple typed search attributes in one statement
+* Support for typed registration of search attributes
+* Support for predefined search attributes in any form of query building
+
+## Proposed solution
+
+### General
+
+* All SDK workflow APIs will now have a mutable search attribute map with typed getters and setters
+  * When a mutation on a search attribute occurs:
+    * If the most recent command is an upsert search attribute, the single value is merged
+    * If the most recent command is not an upsert search attribute, make one
```

Comment:
>  Other approaches are - not merged and preserve the correlation between the call site and the command or flush the entire batch at the end of each "activation".

I don't like either of these other approaches. The correlation per single value is not needed, and flushing at the end is hiding too much, but you do want _some_ kind of merging just because we are no longer providing a way to update multiple in one statement (unless dicussion about.

> If we're going for batching I'd rather defer the flushing to the end to get the most compact command form.

Not looking for most compact, looking for ability to replace upsert w/ multiple synchronous single-value setters and retain same event counts. If I used to upsert 10 keys in a single call (or 1 in a loop without commands interspersed) and I change to this single-value upsert, I don't need 10 events.

> Also, please note that this method will not be history compatible with the older APIs and since we plan to deprecate that older APIs, I'm afraid users might rush to change their code without understanding the implications.

I hope not. Nobody should use new command-causing workflow APIs and expect compatibility. But we can call it out.